### ISSUE TITLE
`bracketOnError` / `onException` are unsafe in the presence of multiple short-circuiting effects

### ISSUE THREAD

**Issue by lexi-lambda** [polysemy-research/polysemy#262]
I’m sorry for not getting back to you about #87, where you asked me if it satisfied my concerns, but it does not. Consider: let’s say I’m interacting with a database, and on exceptions, I want to rollback the current transaction. So I do something like this:

```haskell
withTransaction :: (Member Resource r, Member Trace r) => Sem r a -> Sem r a
withTransaction m =
  bracketOnError
    (trace "beginning transaction")
    (const $ trace "rolling back transaction")
    (const $ m <* trace "committing transaction")
```

If I use this, and an IO exception is thrown, the transaction will be rolled back, as I expect:

```haskell
ghci> runFinal . resourceToIOFinal . embedToFinal @IO . traceToIO $
        withTransaction (embed . IO.throwIO $ IO.ErrorCall "bang")
beginning transaction
rolling back transaction
*** Exception: bang
```

But what if I add an `Error` effect? Disaster strikes!

```haskell
ghci> runFinal . resourceToIOFinal . embedToFinal @IO . traceToIO . runError @String $
        withTransaction (throw "bang")
beginning transaction
Left "bang"
```

Now I have a dangling transaction that is neither committed nor rolled back. Given this, it seems impossible to write exception-safe code in Polysemy, since you cannot guard against all possible early exit paths.

**Comment by isovector** [comment]
Mind trying this against the (pure) `runResource`? I'd expect it to indeed perform the rollback, and if so, the fix should be trivial for `resourceToIOFinal`.

**Comment by lexi-lambda** [comment]
Yes, it does perform the rollback with `runResource` (for `Error`, but not for IO exceptions of course). So I guess the required fix just involves making `resourceToIO`/`resourceToIOFinal` do what `runResource` does in addition to what they already do?

**Comment by isovector** [comment]
Yup --- not a fundamental bug, just an oversight. Thanks for the bug, I'll try to get a PR out for it tonight.

**Comment by isovector** [comment]
Fixed in 1.2.2.0. Thanks again for reporting this!
